File: SpaceDock/endpoints/accounts.py
from datetime import datetime, timedelta
from flask import request
from flask_login import current_user, login_user, logout_user
from SpaceDock.common import with_session
from SpaceDock.config import cfg
from SpaceDock.database import db
from SpaceDock.email import send_confirmation, send_reset
from SpaceDock.formatting import user_info
from SpaceDock.objects import Mod, Role, User
from SpaceDock.routing import route

import bcrypt
import re
import binascii
import os

@route('/api/register', methods=['POST'])
@with_session
def register():
    """
    Registers a new useraccount. Required parameters: email, username, password, repeatPassword. Optional parameters: follow-mod
    """
    if not cfg.getb('registration'):
        return {'error': True, 'reasons': ['Registrations are disabled'], 'codes': ['3010']}, 400

    followMod = request.json.get('follow-mod')
    email = request.json.get('email')
    username = request.json.get('username')
    password = request.json.get('password')
    confirmPassword = request.json.get('repeatPassword')

    errors = ()
    codes = ()
    emailError = check_email_for_registration(email)
    if emailError:
        errors.append(emailError)
        codes.append('4000')

    usernameError = check_username_for_registration(username)
    if usernameError:
        errors.append(usernameError)
        codes.append('4000')

    if not password:
        errors.append('Password is required.')
        codes.append('2515')
    else:
        if password != confirmPassword:
            errors.append('Passwords do not match.')
            codes.append('3005')
        if len(password) < 5:
            errors.append('Your password must be greater than 5 characters.')
            codes.append('2101')
        if len(password) > 256:
            errors.append('We admire your dedication to security, but please use a shorter password.')
            codes.append('2102')

    if len(errors) > 0:
        return {'error': True, 'reasons': errors, 'codes': codes}, 400

    # All valid, let's make them an account
    user = User(username, email, password)
    user.confirmation = binascii.b2a_hex(os.urandom(20)).decode("utf-8")
    db.add(user)
    db.commit() # We do this manually so that we're sure everything's hunky dory before the email leaves
    if followMod:
        send_confirmation(user, followMod)
    else:
        send_confirmation(user)
    return {'error': False, 'count': 1, 'data': user_info(user)}

def check_username_for_registration(username):
    if not username:
        return 'Username is required.'
    if not re.match(r"^[A-Za-z0-9_]+$", username):
        return 'Please only use letters, numbers, and underscores.'
    if len(username) < 3 or len(username) > 24:
        return 'Usernames must be between 3 and 24 characters.'
    if User.query.filter(User.username.ilike(username)).first():
        return 'A user by this name already exists.'
    return None

def check_email_for_registration(email):
    if not email:
        return 'Email is required.'
    if not re.match(r"^[^@]+@[^@]+\.[^@]+$", email):
        return 'Please specify a valid email address.'
    elif User.query.filter(User.email == email).first():
        return 'A user with this email already exists.'
    return None

@route('/api/confirm/<username>/<confirmation>')
@with_session
def confirm(username, confirmation):
    """
    Confirms the user. The confirmation must match what was sent to the user.
    """
    user = User.query.filter(User.username == username).first()
    if not user:
        return {'error': True, 'reasons': ['User does not exist'], 'codes': ['2165']}, 400
    if user.confirmation == None:
        return {'error': True, 'reasons': ['User already confirmed'], 'codes': ['3045']}, 400
    if user.confirmation != confirmation:
        return {'error': True, 'reasons': ['Confirmation does not match'], 'codes': ['2100']}, 400

    user.confirmation = None
    login_user(user)
    user.add_roles(username)
    role = Role.query.filter(Role.name == username).first()
    role.add_abilities('user-edit', 'mods-add', 'packs-add', 'logged-in')
    role.add_param('user-edit', 'userid', user.id)
    role.add_param('mods-add', 'gameshort', '.*')
    role.add_param('packs-add', 'gameshort', '.*')
    db.add(role)
    f = request.args.get('f')
    if f:
        mod = Mod.query.filter(Mod.id == int(f)).first()
        mod.follower_count += 1
        user.following.append(mod)
    return {'error': False}

@route('/api/login', methods=['POST'])
def login():
    """
    Login the user to use additional features. Required fields: username, password
    """
    username = request.json.get('username')
    password = request.json.get('password')
    if not username or not password:
        return {'error': True, 'reasons': ['Missing username or password'], 'codes': ['2515']}, 400
    if current_user:
        return {'error': True, 'reasons': ['You are already logged in'], 'codes': ['3060']}, 400
    user = User.query.filter(User.username.ilike(username)).first()
    if not user:
        return {'error': True, 'reasons': ['Username or password is incorrect'], 'codes': ['2175']}, 400
    if not bcrypt.hashpw(password.encode('utf-8'), user.password.encode('utf-8')) == user.password.encode('utf-8'):
        return {'error': True, 'reasons': ['Username or password is incorrect'], 'codes': ['2175']}, 400
    if user.confirmation == '' and user.confirmation == None:
        return {'error': True, 'reasons': ['User is not confirmed'], 'codes': ['3055']}, 400
    login_user(user)
    return {'error': False}

@route('/api/logout')
def logout():
    """
    Closes the session and logs the user out
    """
    if not current_user:
        return {'error': True, 'reasons': ['You are not logged in. Logging out now would be a bit difficult, right?'], 'codes': ['3070']}, 400 # Thats my daily portion of irony
    logout_user()
    return {'error': False}

@route('/api/reset', methods=['POST'])
@with_session
def forgot_password():
    """
    Sends you a confirmation key for resetting your password in case you forgot it. Required fields: email
    """
    email = request.json.get('email')
    if not email:
        return {'error': True, 'reasons': ['No email address'], 'codes': ['2520']}, 400
    user = User.query.filter(User.email == email).first()
    if not user:
        return {'error': True, 'reasons': ['No user for provided email address'], 'codes': ['2115']}, 400
    user.passwordReset = binascii.b2a_hex(os.urandom(20)).decode("utf-8")
    user.passwordResetExpiry = datetime.now() + timedelta(days=1)
    send_reset(user)
    return {'error': False}

@route('/api/reset/<username>/<confirmation>', methods=['POST'])
@with_session
def reset_password(username, confirmation):
    """
    Allows you to reset your password with the key you got through /api/reset. Required parameters: password, password2
    """
    user = User.query.filter(User.username == username).first()
    if not user:
        return {'error': True, 'reasons': ['Username is incorrect'], 'codes': ['2170']}, 400

    if user.passwordResetExpiry == None or user.passwordResetExpiry < datetime.now():
        return {'error': True, 'reasons': ['Password reset invalid'], 'codes': ['3000']}, 400
    if user.passwordReset != confirmation:
        return {'error': True, 'reasons': ['Password reset invalid'], 'codes': ['3000']}, 400
    password = request.json.get('password')
    password2 = request.json.get('password2')
    if not password or not password2:
        return {'error': True, 'reasons': ['Passwords not provided'], 'codes': ['2525']}, 400
    if password != password2:
        return {'error': True, 'reasons': ['Passwords do not match'], 'codes': ['3005']}, 400
    user.set_password(password)
    user.passwordReset = None
    user.passwordResetExpiry = None
    return {'error': False}

File: src/SpaceDock/routes/accounts.go
/*
 SpaceDock Backend
 API Backend for the SpaceDock infrastructure to host modfiles for various games

 SpaceDock Backend is licensed under the Terms of the MIT License.
 Copyright (c) 2017 Dorian Stoll (ThomasKerman/TMSP), RockyTV
*/

package routes

import (
    "SpaceDock"
    "SpaceDock/middleware"
    "SpaceDock/objects"
    "SpaceDock/utils"
    "github.com/jameskeane/bcrypt"
    "gopkg.in/kataras/iris.v6"
    "regexp"
    "strconv"
    "time"
    "github.com/spf13/cast"
)

/*
 Registers the routes for the account management
 */
func AccountsRegister() {
    Register(POST, "/api/register", register)
    Register(GET, "/api/confirm/:confirmation", confirm) // Maybe switch to POST too?
    Register(POST, "/api/login", login)
    Register(POST, "/api/logout", logout)
    Register(POST, "/api/reset", reset)
    Register(POST, "/api/reset/:username/:confirmation", resetConfirm)
}

/*
 Path: /api/register
 Method: POST
 Description: Creates a new useraccount
 */
func register(ctx *iris.Context) {
    // Check if registration is allowed
    if !SpaceDock.Settings.Registration {
        utils.WriteJSON(ctx, iris.StatusForbidden, utils.Error("Registration is disabled").Code(3010))
        return
    }

    // Grab parameters from the JSON
    followMod := cast.ToString(utils.GetJSON(ctx,"follow-mod"))
    email := cast.ToString(utils.GetJSON(ctx,"email"))
    username := cast.ToString(utils.GetJSON(ctx,"username"))
    password := cast.ToString(utils.GetJSON(ctx,"password"))
    confirmPassword := cast.ToString(utils.GetJSON(ctx,"repeatPassword"))
    data := cast.ToStringMap(utils.GetJSON(ctx,"userdata"))
    check := cast.ToString(utils.GetJSON(ctx,"check"))

    var errors []string
    var codes []int
    emailError := checkEmailForRegistration(email)
    if emailError != "" {
        errors = append(errors, emailError)
        codes = append(codes, 4000)
        if check == "email" {
            utils.WriteJSON(ctx, iris.StatusOK, utils.Error(emailError).Code(4000))
            return
        }
    }

    usernameError := checkUsernameForRegistration(username)
    if usernameError != "" {
        errors = append(errors, usernameError)
        codes = append(codes, 4000)
        if check == "username" {
            utils.WriteJSON(ctx, iris.StatusOK, utils.Error(usernameError).Code(4000))
            return
        }
    }

    if password == "" {
        errors = append(errors, "Password is required")
        codes = append(codes, 2515)
        if check == "password" {
            utils.WriteJSON(ctx, iris.StatusOK, utils.Error("Password is required").Code(2515))
            return
        }
    } else {
        if password != confirmPassword {
            errors = append(errors, "Passwords do not match")
            codes = append(codes, 3005)
            if check == "password" {
                utils.WriteJSON(ctx, iris.StatusOK, utils.Error("Passwords do not match").Code(3005))
                return
            }
        }
        if len(password) < 5 {
            errors = append(errors, "Your password must be greater than 5 characters")
            codes = append(codes, 2101)
            if check == "password" {
                utils.WriteJSON(ctx, iris.StatusOK, utils.Error("Your password must be greater than 5 characters").Code(2101))
                return
            }
        }
        if len(password) > 256 {
            errors = append(errors, "We admire your dedication to security, but please use a shorter password")
            codes = append(codes, 2102)
            if check == "password" {
                utils.WriteJSON(ctx, iris.StatusOK, utils.Error("We admire your dedication to security, but please use a shorter password").Code(2102))
                return
            }
        }
    }

    if check == "email" || check == "password" || check == "username" {
        utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false})
        return
    }
    if len(errors) > 0 {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error(errors...).Code(codes...))
        return
    }

    // Everything is valid, make them an account
    user := objects.NewUser(username, email, password)
    user.Confirmation,_ = utils.RandomHex(20)

    // Edit user
    if data != nil {
        code := utils.EditObject(&user, data)
        if code == 3 {
            utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("The value you submitted is invalid").Code(2180))
            return
        } else if code == 2 {
            utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("You tried to edit a value that doesn't exist.").Code(3090))
            return
        } else if code == 1 {
            utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("You tried to edit a value that is marked as read-only.").Code(3095))
            return
        }
    }

    SpaceDock.Database.Save(user)
    utils.SendConfirmation(user.Confirmation, user.Username, user.Email, followMod)

    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false, "count": 1, "data": user})
}

func checkUsernameForRegistration(username string) string {
    if username == "" {
        return "Username is required"
    }
    r,_ := regexp.Compile("^[A-Za-z0-9_]+$")
    var user objects.User
    if !r.MatchString(username) {
        return "Please only use letters, numbers, and underscores"
    }
    if len(username) < 3 || len(username) > 24 {
        return "Usernames must be between 3 and 24 characters"
    }
    if SpaceDock.Database.Where("username = ?", username).First(&user); user.Username != "" {
        return "A user by this name already exists"
    }
    return ""

}

func checkEmailForRegistration(email string) string {
    if email == "" {
        return "Email is required"
    }
    r,_ := regexp.Compile("^[^@]+@[^@]+.[^@]+$")
    var user objects.User
    if !r.MatchString(email) {
        return "Please specify a valid email address."
    } else if SpaceDock.Database.Where("email = ?", email).First(&user); user.Username != "" {
        return "A user with this email already exists."
    }
    return ""
}

/*
 Path: /api/confirm/:confirmation
 Method: GET
 Description: Confirms a newly created useraccount using a random text sequence
 */
func confirm(ctx *iris.Context) {
    // Grab the confirmation sequence
    confirmation := ctx.GetString("confirmation")

    // Try to get a valid user account
    var user *objects.User
    SpaceDock.Database.Where("confirmation = ?", confirmation).First(user)
    if user.Confirmation != confirmation {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("User does not exist or it is already confirmed. Did you mistype the confirmation?").Code(2165))
        return
    }

    // Everything is valid
    user.Confirmation = ""
    middleware.LoginUser(ctx, user)
    role := user.AddRole(user.Username)
    role.AddAbility("user-edit")
    role.AddAbility("mods-add")
    role.AddAbility("packs-add")
    role.AddAbility("logged-in")
    role.AddParam("user-edit", "userid", strconv.Itoa(int(user.ID)))
    role.AddParam("mods-add", "gameshort", ".*")
    role.AddParam("packs-add", "gameshort", ".*")
    SpaceDock.Database.Save(&role)
    SpaceDock.Database.Save(&user)

    // Follow Mod

    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false})
}

/*
 Path: /api/login
 Method: POST
 */
func login(ctx *iris.Context) {
    // Grab information
    username := cast.ToString(utils.GetJSON(ctx, "username"))
    password := cast.ToString(utils.GetJSON(ctx, "password"))

    // Check if the values are valid
    if username == "" || password == "" {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("Missing username or password").Code(2515))
        return
    }
    if middleware.CurrentUser(ctx) != nil {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("You are already logged in").Code(3060))
        return
    }
    var user objects.User
    SpaceDock.Database.Where("username = ?", username).First(&user)
    if user.Username != username {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("Username or password is incorrect").Code(2175))
        return
    }
    if s,_ := bcrypt.Hash(password, user.Password); s != user.Password {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("Username or password is incorrect").Code(2175))
        return
    }
    if user.Confirmation != "" {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("User is not confirmed").Code(3055))
        return
    }
    middleware.LoginUser(ctx, &user)
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false, "count": 1, "data": user})
}

/*
 Path: /api/logout
 Method: GET
 */
func logout(ctx *iris.Context) {
    // Check if a user is logged in
    user := middleware.CurrentUser(ctx)
    if user == nil {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("You are not logged in. Logging out now would be a bit difficult, right?").Code(3070))
        return
    }
    middleware.LogoutUser(ctx)
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false})
}

/*
 Path: /api/reset
 Method: POST
 */
func reset(ctx *iris.Context) {
    // Get the email
    email := cast.ToString(utils.GetJSON(ctx, "email"))

    // Check if the values are valid
    if email == "" {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("No email address").Code(2520))
        return
    }
    var user objects.User
    SpaceDock.Database.Where("email = ?", email).First(&user)
    if user.Email != email {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("No user for provided email address").Code(2115))
        return
    }
    user.PasswordReset,_ = utils.RandomHex(20)
    user.PasswordResetExpiry = time.Now().Add(time.Hour * 24)
    SpaceDock.Database.Save(&user)
    utils.SendReset(user.Username, user.PasswordReset, user.Email)
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false})
}

/*
 Path: /api/reset/:username/:confirmation
 Method: POST
 */
func resetConfirm(ctx *iris.Context) {
    username := ctx.GetString("username")
    confirmation := ctx.GetString("confirmation")
    var user objects.User
    SpaceDock.Database.Where("username = ?", username).First(&user)
    if user.Username != username {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("Username is incorrect").Code(2170))
        return
    }
    if user.PasswordReset == "" || user.PasswordResetExpiry.Before(time.Now()) || user.PasswordReset != confirmation {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("Password reset invalid").Code(3000))
        return
    }
    password := utils.GetJSON(ctx, "password").(string)
    password2 := utils.GetJSON(ctx, "password2").(string)
    if password == "" || password2 == "" {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("Passwords not provided").Code(2525))
        return
    }
    if password != password2 {
        utils.WriteJSON(ctx, iris.StatusBadRequest, utils.Error("Passwords do not match").Code(3005))
        return
    }
    user.SetPassword(password)
    user.PasswordReset = ""
    user.PasswordResetExpiry = time.Now()
    SpaceDock.Database.Save(&user)
    if middleware.CurrentUser(ctx) != nil {
        middleware.LogoutUser(ctx)
    }
    utils.WriteJSON(ctx, iris.StatusOK, iris.Map{"error": false})
}
